drivers/include/periph/hwcrypto.h

```diff
@@ -0,0 +1,323 @@
+ * internal peripheral state (e.g. load the key, keep the digest interally),
+ * therefore must be used to control mutual exclusive access.
+ *
+ * @{
+ * @file
+ * @brief       Low-level hardware crypto driver interface definitions
+ *
+ * @author      Bas Stottelaar <basstottelaar@gmail.com>
+ */
+
+#ifndef HWCRYPTO_H
+#define HWCRYPTO_H
+
+#include <stdint.h>
+
+#include "periph_cpu.h"
+#include "periph_conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* guard file in case no HWCRYPTO device is defined */
+#if HWCRYPTO_NUMOF
+
+/**
+ * @brief   Supported ciphers.
+ *
+ * Define HAVE_HWCRYPTO_xxx in your CPUs periph_cpu.h if needed.
+ *
+ * Extend at will.
+ */
+typedef enum {
+#ifdef HAVE_HWCRYPTO_AES128
+    HWCRYPTO_AES128,                  /**< AES128 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_AES256
+    HWCRYPTO_AES256,                  /**< AES256 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_DES
+    HWCRYPTO_DES,                     /**< DES cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_3DES
+    HWCRYPTO_3DES,                    /**< 3DES cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_TWOFISH
+    HWCRYPTO_TWOFISH,                 /**< Two-fish cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA512
+    HWCRYPTO_RSA512,                  /**< RSA512 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA1024
+    HWCRYPTO_RSA1024,                 /**< RSA1024 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_RSA2048
+    HWCRYPTO_RSA2048,                 /**< RSA2048 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC128
+    HWCRYPTO_ECC128,                  /**< ECC128 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC192
+    HWCRYPTO_ECC192,                  /**< ECC192 cipher */
+#endif
+#ifdef HAVE_HWCRYPTO_ECC256
+    HWCRYPTO_ECC256,                  /**< ECC256 cipher */
+#endif
+    HWCRYPT_CIPHER_NULL,
+} hwcrypto_cipher_t;
+
+/**
+ * @brief   Supported hashing algorithms.
+ *
+ * Define HAVE_HWCRYPTO_xxx in your CPUs periph_cpu.h if needed.
+ *
+ * Extend at will.
+ */
+typedef enum {
+#ifdef HAVE_HWCRYPTO_MD4
+    HWCRYPTO_MD4,                     /**< MD4 hashing algorithm */
+#endif
+#ifdef HAVE_HWCRYPTO_MD5
+    HWCRYPTO_MD5,                     /**< MD5 hashing algorithm */
+#endif
+#ifdef HAVE_HWCRYPTO_SHA1
+    HWCRYPTO_SHA1,                    /**< SHA1 hashing algorithm */
+#endif
+#ifdef HAVE_HWCRYPTO_SHA256
+    HWCRYPTO_SHA256,                  /**< SHA256 hashing algorithm */
+#endif
+    HWCRYPT_HASH_NULL
+} hwcrypto_hash_t;
+
+/**
+ * @brief   Default cipher context type for ciphers with keys up to 256 bits.
+ *
+ * The context should store the selected cipher. If needed, the context should
+ * provide a buffer to store a key.
+ *
+ * For performance reasons, the default context type aligns the key.
+ */
+#ifndef HAVE_HWCRYPTO_CIPHER_CONTEXT_T
+typedef struct {
+    hwcrypto_cipher_t cipher;                       /**< selected cipher method */
+    uint8_t __attribute__ ((aligned)) key[32];      /**< context for up to 256-bit key */
```

Comment:
The hardware modules may have different alignment requirements than the CPU
I guess, at least if they are fed via DMA so the CPU doesn't handle the
data transfer.
On May 11, 2016 5:23 PM, "Bas Stottelaar" <notifications@github.com> wrote:

> In drivers/include/periph/hwcrypto.h
> <https://github.com/RIOT-OS/RIOT/pull/5429#discussion_r62879771>:
>
> > +#endif
> > +    HWCRYPT_HASH_NULL
> > +} hwcrypto_hash_t;
> > +
> > +/**
> > + * @brief   Default cipher context type for ciphers with keys up to 256 bits.
> > + *
> > + * The context should store the selected cipher. If needed, the context should
> > + * provide a buffer to store a key.
> > + *
> > + * For performance reasons, the default context type aligns the key.
> > + */
> > +#ifndef HAVE_HWCRYPTO_CIPHER_CONTEXT_T
> > +typedef struct {
> > +    hwcrypto_cipher_t cipher;                       /**< selected cipher method */
> > +    uint8_t __attribute__ ((aligned)) key[32];      /**< context for up to 256-bit key */
>
> According to this
> <https://gcc.gnu.org/onlinedocs/gcc-4.0.2/gcc/Type-Attributes.html>:
>
> "Alternatively, you can leave out the alignment factor and just ask the
> compiler to align a type to the maximum useful alignment for the target
> machine you are compiling for."
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/5429/files/7b05094eb974d7212663869bc0078a8f357ff279#r62879771>
>